python/pyspark/sql/catalog.py

```diff
@@ -255,26 +255,67 @@ def registerFunction(self, name, f, returnType=StringType()):
         >>> _ = spark.udf.register("stringLengthInt", len, IntegerType())
         >>> spark.sql("SELECT stringLengthInt('test')").collect()
         [Row(stringLengthInt(test)=4)]
+        """
+
+        # This is to check whether the input function is a wrapped/native UserDefinedFunction
+        if hasattr(f, 'asNondeterministic'):
+            raise ValueError("Please use registerUDF for registering UDF. The expected function of "
+                             "registerFunction is a Python function (including lambda function)")
+        udf = UserDefinedFunction(f, returnType=returnType, name=name,
+                                  evalType=PythonEvalType.SQL_BATCHED_UDF)
+        self._jsparkSession.udf().registerPython(name, udf._judf)
+        return udf._wrapped()
+
+    @ignore_unicode_prefix
+    @since(2.3)
+    def registerUDF(self, name, f):
+        """Registers a :class:`UserDefinedFunction`. The registered UDF can be used in SQL
+        statement.
+
+        :param name: name of the UDF
+        :param f: a wrapped/native UserDefinedFunction. The UDF can be either row-at-a-time or
```

Comment:
We should probably standardize the how to describe the wrapped function object returned by `udf` and `pandas_udf` in param docstring. Currently it's being described differently:

https://github.com/apache/spark/blob/master/python/pyspark/sql/group.py#L215

Reply:
I don't have a strong opinion what to describe them, but I think it's easy for user to understand if the description is consistent.

Reply:
> :param udf: A function object returned by :meth:`pyspark.sql.functions.pandas_udf`

This is actually not accurate. We have another way to define the scalar vectorized UDF.
```
>>> @pandas_udf("integer", PandasUDFType.SCALAR)  # doctest: +SKIP
... def add_one(x):
...     return x + 1
...
```

Reply:
Yeah. I am fine with the doc here. But similar to https://github.com/apache/spark/pull/20217#discussion_r160979747 we should probably standardize the description.

Reply:
We can do it as a separate PR.

Reply:
SGTM.